rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,675 @@
+
+   - Opt-in warnings
+     - (-) Can't really make breaking changes as people won't be warned ahead of time
+   - No warnings, just errors
+     - (-) Doesn't offer a transition window to users
+     - (+) Easier to implement
+
+   </details>
+
+1. A the end of the evaluation, print statistics and explanations.
+   The specifics of displaying warnings and errors is up to implementation, but should include the symbolic name of the langauge construct in question.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Show the current Nix language version
+
+```console
+nix --language-version
+7
+```
+
+```console
+nix --supported-language-versions
+6
+7
+```
+
+## Version interoperability
+
+### Versioned built-ins can be passed across file boundaries
+
+```nix
+# a.nix
+version 6;
+builtins.langVersion
+```
+
+```nix
+# b.nix
+version 7;
+[ (import ./a.6.nix) builtins.langVersion ]
+```
+
+```console
+$ nix-instantiate --eval b.nix
+[ 6 7 ]
+```
+
+### Expressions are not forward compatible
+
+```nix
+# a.nix
+version 6;
+import ./b.nix
+```
+
+```nix
+# b.nix
+version 7;
+builtins.langVersion
+```
+
+```console
+$ nix-instantiate --eval a.nix
+error: unsupported Nix language version 7
+```
+
+
+### Best-effort interoperability
+
+```nix
+# a.nix
+version 6;
+{ increment }:
+increment 1
+```
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { increment = x: x + 1 }
+```
+
+Since `increment` written in version 7 carries its own implementation with it, forcing it within an expression written in version 6 just works:
+
+```console
+$ nix-instantiate --eval b.nix
+2
+```
+
+### Pathological example
+
+Usually existing code will be interacted with by calling functions.
+When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.
+
+```nix
+# a.nix
+version 6;
+{ value }:
+value + 1
+```
+
+Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { value = %5 + 7i%; }
+```
+
+```console
+$ nix-instantiate --eval bnix
+error: unsupported value type `complex` at built-in operator `+`
+```
+
+## Deprecation warnings
+
+```
+nix eval --json ./test.nix
+warning: URL literals are deprecated (url-literal)
+         please replace this with a string: "https://nixos.org"
+
+       at test.nix:1:1:
+
+            1| https://nixos.org
+             | ^
+
+"https://nixos.org"
+
+warning: The following deprecated features were used:
+  - url-literal (httsp://nixos.org), 1 time
+
+  Add `--lang-warn-verbose` to show all occurrences
+  Use `--lang-no-warn=url-literal` to disable this warning.
+  Use `--lang-error=url-literal` to issue errors instead of warnings.
+```
+
+## Drawbacks
+
+Allowing multiple language versions to coexist complicates implementation of valuators and support tooling, and makes comprehensive test coverage harder.
```

Comment:
```suggestion
Allowing multiple language versions to coexist complicates implementation of evaluators and support tooling, and makes comprehensive test coverage harder.
All else being equal, it may increase maintenance burden and the likelihood of introducing bugs.
```